pkg/plugins/connection/ssh.go

```diff
@@ -30,30 +30,46 @@ func (s SSH) String() string {
 	return "  - connection opened"
 }
 
+// ClientConfigFn is an interface that allows users to implement their own SSH auth mechanisms
+type ClientConfigFn func(*gornir.Host, gornir.Logger) (*ssh.ClientConfig, error)
+
 // SSHOpen is a Connection plugin that opens a connection with a device
 type SSHOpen struct {
-	Meta *gornir.TaskMetadata // Task metadata
+	Meta           *gornir.TaskMetadata // Task metadata
+	ClientConfigFn ClientConfigFn       // SSH client configuration
 }
 
 // Metadata returns the task metadata
 func (t *SSHOpen) Metadata() *gornir.TaskMetadata {
 	return t.Meta
 }
 
-// Run implements gornir.Task interface
-func (t *SSHOpen) Run(ctx context.Context, logger gornir.Logger, host *gornir.Host) (gornir.TaskInstanceResult, error) {
-	sshConfig := &ssh.ClientConfig{
+// defaultSSHClientConfig implements ClientConfigFn
+func defaultSSHClientConfig(host *gornir.Host, logger gornir.Logger) (*ssh.ClientConfig, error) {
+	return &ssh.ClientConfig{
 		User: host.Username,
 		Auth: []ssh.AuthMethod{
 			ssh.Password(host.Password),
 		},
 		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
-	} // #nosec
+	}, nil // #nosec
+}
+
+// Run implements gornir.Task interface
+func (t *SSHOpen) Run(ctx context.Context, logger gornir.Logger, host *gornir.Host) (gornir.TaskInstanceResult, error) {
+	var clientConfigFn ClientConfigFn = defaultSSHClientConfig
```

Comment:
a bit nit picky but for consistency I'd use the form  `clientConfigFn := defaultSSHClientConfig`